pkg/streamingpromql/operators/binary_operation.go

```diff
@@ -404,62 +405,111 @@ func (b *BinaryOperation) mergeOneSide(data []types.InstantVectorSeriesData, sou
 	}
 
 	slices.SortFunc(data, func(a, b types.InstantVectorSeriesData) int {
```

Comment:
If I'm understanding correctly you're suggesting:

1. Merge all the floats
2. Merge all the histograms
3. Iterate over the points again and check for conflicts between floats and histograms

Instead of the extra iteration at point 3, what if we break the contract that an FPoint might exist at the same time as an HPoint? Then in `computeResult` (or more specifically `InstantVectorSeriesDataIterator`) we raise an error where both exist?

Reply:
> Instead of the extra iteration at point 3, what if we break the contract that an FPoint might exist at the same time as an HPoint? Then in `computeResult` (or more specifically `InstantVectorSeriesDataIterator`) we raise an error where both exist?

I think that will make `computeResult` more complex, and means the responsibility of checking for conflicts is split across multiple methods. 

If your concern is performance, we should be able to skip the third step most of the time, as we only need to do it if we have both floats and histograms for one side.

Reply:
My concern is performance as it can be a lot of points to iterate over. The chances of it happening in correctly formed queries should be low, but I can also see it being not too difficult to have a selector that accidentally selects multiple types of metrics.

I'll have it separate for now but leave a note that we could revisit it. Another option, for example, would be to track the timestamps that we have a point for already etc.

Reply:
I've started to implement this, but the problem this approach has is that we can't tell which series the conflict comes from. How important is it that we emit that?

Reply:
> My concern is performance as it can be a lot of points to iterate over. 

We still have to do roughly the same amount of work either way: either we do all the checks in one loop, or we do them in separate loops. The cost of an extra `for` loop is relatively small compared to the cost of the conflict detection logic.

> I've started to implement this, but the problem this approach has is that we can't tell which series the conflict comes from. How important is it that we emit that?

I think it would be fine to emit a generic message based on the output group name for now (eg. `found both float and histogram samples for the match group {foo="blah"} on the left side of the operation at timestamp 123`). We can add the series that the conflict comes from as a future improvement.